eureka-client/src/main/java/com/netflix/discovery/DiscoveryClient.java

```diff
@@ -197,12 +198,32 @@
         private Provider<HealthCheckHandler> healthCheckHandlerProvider;
     }
 
+    /**
+     * Assumes applicationInfoManager is already initialized
+     *
+     * @deprecated use constructor that takes ApplicationInfoManager instead of InstanceInfo directly
+     */
+    @Deprecated
     public DiscoveryClient(InstanceInfo myInfo, EurekaClientConfig config) {
         this(myInfo, config, null);
     }
 
+    /**
+     * Assumes applicationInfoManager is already initialized
+     *
+     * @deprecated use constructor that takes ApplicationInfoManager instead of InstanceInfo directly
+     */
+    @Deprecated
     public DiscoveryClient(InstanceInfo myInfo, EurekaClientConfig config, DiscoveryClientOptionalArgs args) {
-        this(myInfo, config, args, new Provider<BackupRegistry>() {
+        this(ApplicationInfoManager.getInstance(), config, args);
+    }
+
+    public DiscoveryClient(ApplicationInfoManager applicationInfoManager, EurekaClientConfig config) {
+        this(applicationInfoManager, config, null);
+    }
+
+    public DiscoveryClient(ApplicationInfoManager applicationInfoManager, EurekaClientConfig config, DiscoveryClientOptionalArgs args) {
+        this(applicationInfoManager, config, args, new Provider<BackupRegistry>() {
```

Comment:
Is BackupRegistry supposed to be a singleton?  If so, then cache the reference in get().